# **CSCE 5218 / CSCE 4930 Deep Learning**

# **HW1a The Perceptron** (20 pt)


In [140]:
# Take a peek at the datasets
!head train.dat
!head test.dat

A1	A2	A3	A4	A5	A6	A7	A8	A9	A10	A11	A12	A13	
1	1	0	0	0	0	0	0	1	1	0	0	1	0
0	0	1	1	0	1	1	0	0	0	0	0	1	0
0	1	0	1	1	0	1	0	1	1	1	0	1	1
0	0	1	0	0	1	0	1	0	1	1	1	1	0
0	1	0	0	0	0	0	1	1	1	1	1	1	0
0	1	1	1	0	0	0	1	0	1	1	0	1	1
0	1	1	0	0	0	1	0	0	0	0	0	1	0
0	0	0	1	1	0	1	1	1	0	0	0	1	0
0	0	0	0	0	0	1	0	1	0	1	0	1	0
A1	A2	A3	A4	A5	A6	A7	A8	A9	A10	A11	A12	A13
1	1	1	1	0	0	1	1	0	0	0	1	1	0
0	0	0	1	0	0	1	1	0	1	0	0	1	0
0	1	1	1	0	1	1	1	1	0	0	0	1	0
0	1	1	0	1	0	1	1	1	0	1	0	1	0
0	1	0	0	0	1	0	1	0	1	0	0	1	0
0	1	1	0	0	1	1	1	1	1	1	0	1	0
0	1	1	1	0	0	1	1	0	0	0	1	1	0
0	1	0	0	1	0	0	1	1	0	1	1	1	0
1	1	1	1	0	0	1	1	0	0	0	0	1	0


### Build the Perceptron Model

You will need to complete some of the function definitions below.  DO NOT import any other libraries to complete this. 

In [180]:
import math
import itertools
import re


# Corpus reader, all columns but the last one are coordinates;
#   the last column is the label
def read_data(file_name):
    f = open(file_name, 'r')

    data = []
    # Discard header line
    f.readline()
    for instance in f.readlines():
        if not re.search('\t', instance): continue
        instance = list(map(int, instance.strip().split('\t')))
        # Add a dummy input so that w0 becomes the bias
        w0 = 0.5
        instance = [-1] + instance
        data += [instance]
    return data


def dot_product(array1, array2):
    #TODO: Return dot product of array 1 and array 2
    weighted_sum = 0 
    for i in range(len(array1)):
        weighted_sum += array1[i]*array2[i]
    return weighted_sum


def sigmoid(x):
    #TODO: Return outpout of sigmoid function on x
    return 1.0/(1.0+math.exp(-x)) 


# The output of the model, which for the perceptron is 
# the sigmoid function applied to the dot product of 
# the instance and the weights
def output(weight, instance):
    #TODO: return the output of the model 
    return sigmoid(dot_product(weights, instance))

# Predict the label of an instance; this is the definition of the perceptron
# you should output 1 if the output is >= 0.5 else output 0
def predict(weights, instance):
    bias = 0.5
    #TODO: return the prediction of the model
    if output(weights, instance) >=bias:
        return 1
    else: 
        return 0


# Accuracy = percent of correct predictions
def get_accuracy(weights, instances):
    # You do not to write code like this, but get used to it
    correct = sum([1 if predict(weights, instance) == instance[-1] else 0
                   for instance in instances])
    return correct * 100 / len(instances)


# Train a perceptron with instances and hyperparameters:
#       lr (learning rate) 
#       epochs
# The implementation comes from the definition of the perceptron

# Training consists on fitting the parameters which are the weights
# that's the only thing training is responsible to fit
# (recall that w0 is the bias, and w1..wn are the weights for each coordinate)
# Hyperparameters (lr and epochs) are given to the training algorithm
# We are updating weights in the opposite direction of the gradient of the error,
# so with a "decent" lr we are guaranteed to reduce the error after each iteration.
def train_perceptron(instances, lr, epochs):

    #TODO: name this step
    weights = [0] * (len(instances[0])-1)

    for _ in range(epochs):
        for instance in instances_tr:
            #TODO: name these steps
            in_value = dot_product(weights, instance)
            output = sigmoid(in_value)
            error = instance[-1] - output
            #TODO: name these steps
            for i in range(0, len(weights)):
                weights[i] += lr * error * output * (1-output) * instance[i]

    return weights

## Run it

In [181]:
instances_tr = read_data('train.dat')
instances_te = read_data('test.dat')

In [182]:
print(len(instances_tr))

400


In [183]:
print(len(instances_te))

100


In [185]:
lr = 0.005
epochs = 5
weights = train_perceptron(instances=instances_tr, lr = lr, epochs=epochs)
accuracy = get_accuracy(weights, instances_te)
print(f"#tr: {len(instances_tr):3}, epochs: {epochs:3}, learning rate: {lr:.3f}; "
      f"Accuracy (test, {len(instances_te)} instances): {accuracy:.1f}")

#tr: 400, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0


### TEST CASE

In [187]:
lr = 0.005
epochs = 100
weights = train_perceptron(instances=instances_tr, lr = lr, epochs=epochs)
accuracy = get_accuracy(weights, instances_te)
print(f"#tr: {len(instances_tr):3}, epochs: {epochs:3}, learning rate: {lr:.3f}; "
      f"Accuracy (test, {len(instances_te)} instances): {accuracy:.1f}")

#tr: 400, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 77.0


## Questions

Answer the following questions. Include your implementation and the output for each question.



### Question 1

In `train_perceptron(instances, lr, epochs)`, we have the follosing code:
```
in_value = dot_product(weights, instance)
output = sigmoid(in_value)
error = instance[-1] - output
```

Why don't we have the following code snippet instead?
```
output = predict(weights, instance)
error = instance[-1] - output
```

#### TODO Add your answer here (text only)

**Ans.** We use the above code snippet because first we initialize the input value to the weighted sum of the inputs and their respective weights. Then we pass this input value through an activation function, in our case it is sigmoid function. This sigmoid function is really crucial in deciding the accuracy of the deep learning model. Sigmoid activation function limits the final output between 0 and 1 and this helps to mitigate the vanishing gradients problem where the value of the update grows exponentially and can lead to certain issues such as the vanishing gradient problem. Another most important function of this code snippet is that it adds non linearity to the neural network. Although sigmoid is computationally expensive, we use it because it works fine and gives perfect results especially for binary classification problems.

On the other hand, if we do not use activation functions, and we directly predict the output based on the weights and input instances, the network cannot train itself for higher parameters and also can lead to variety of problems that involves with linearity of the neural network.

Hence by passing the output through an activation function, we can tweak the hyperparamters and update our initial guess of the weights for better accuracy of the model.



### Question 2
Train the perceptron with the following hyperparameters and calculate the accuracy with the test dataset.

```
tr_percent = [5, 10, 25, 50, 75, 100] # percent of the training dataset to train with
num_epochs = [5, 10, 20, 50, 100]              # number of epochs
lr = [0.005, 0.01, 0.05]              # learning rate
```

TODO: Write your code below and include the output at the end of each training loop (NOT AFTER EACH EPOCH)
of your code.The output should look like the following:
```
# tr:  20, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
# tr:  20, epochs:  10, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
# tr:  20, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
[and so on for all the combinations]
```
You will get different results with different hyperparameters.

#### TODO Add your answer here (code and output in the format above) 


In [193]:
instances_tr = read_data("train.dat")
instances_te = read_data("test.dat")
tr_percent = [5, 10, 25, 50, 75, 100] # percent of the training dataset to train with
num_epochs = [5, 10, 20, 50, 100]     # number of epochs
lr_array = [0.005, 0.01, 0.05]        # learning rate

for lr in lr_array:
    for tr_size in tr_percent:
        for epochs in num_epochs:
            size =  round(len(instances_tr)*tr_size/100)
            pre_instances = instances_tr[0:size]
            weights = train_perceptron(pre_instances, lr, epochs)
            accuracy = get_accuracy(weights, instances_te)
            print(f"#tr: {len(pre_instances):0}, epochs: {epochs:3}, learning rate: {lr:.3f}; "
            f"Accuracy (test, {len(instances_te)} instances): {accuracy:.1f}")

#tr: 20, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 20, epochs:  10, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 20, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 69.0
#tr: 20, epochs:  50, learning rate: 0.005; Accuracy (test, 100 instances): 73.0
#tr: 20, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 77.0
#tr: 40, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 40, epochs:  10, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 40, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 69.0
#tr: 40, epochs:  50, learning rate: 0.005; Accuracy (test, 100 instances): 73.0
#tr: 40, epochs: 100, learning rate: 0.005; Accuracy (test, 100 instances): 77.0
#tr: 100, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 100, epochs:  10, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
#tr: 100, epochs:  20, lea

### Question 3
Write a couple paragraphs interpreting the results with all the combinations of hyperparameters. Drawing a plot will probably help you make a point. In particular, answer the following:
- A. Do you need to train with all the training dataset to get the highest accuracy with the test dataset?
- B. How do you justify that training the second run obtains worse accuracy than the first one (despite the second one uses more training data)?
   ```
#tr: 100, epochs:  20, learning rate: 0.050; Accuracy (test, 100 instances): 71.0
#tr: 200, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
```
- C. Can you get higher accuracy with additional hyperparameters (higher than `80.0`)?
- D. Is it always worth training for more epochs (while keeping all other hyperparameters fixed)?

#### TODO: Add your answer here (code and text)



From the above interpretation of the results that includes the combinations of all the hyperparameters, we can observe that in the first iteration, we set the learning rate to 0.005 and we can observe that with each update to the training samples and epochs, there is little increase in the overall accuracy of the network. It barely increased from 68% to 70% at the end of training. 
On the other hand, when we increase the learning rate to 0.010, thats when we see a gradual increase in the accuracy of the model.Now as we increase the learning rate, the model quickly converges to its local minima. As it is a simple one neuron network, the task at hand becomes more simple with increased learning rate. It takes less time for the model to learn and train itself.

**Q1.**  Do you need to train with all the training dataset to get the highest accuracy with the test dataset?

**Ans.** Yes it is better to use all training data to train the model to achieve highest accuracy with the test dataset. This is because if we have training data and we dont use it for training, it is similar to having no data at all and we can lose some important information that might affect the accuracy of the model while training it. 

**Q2.** How do you justify that training the second run obtains worse accuracy than the first one (despite the second one uses more training data)?

**Ans** In general if the learning rate is small, it requires more number of epochs to effectively update the weights and train the model. If the learning rate is large, we need less number of epochs to converge the model to its local minima. If we see the above given data, for 20 epochs each, for training sample of 100 samples, the learning rate is 0.050 while for training samples 200, learning rate is 0.005. In ideal scenario, the accuracy of model with  learning rate 0.005 should be highest, but here we have more number of samples to train the model at same number of epochs which leads to decrease in the accuracy in the second run as compared to the first one. 

**Q3.** Can you get higher accuracy with additional hyperparameters (higher than 80.0)?

**Ans.** Accuracy can be increased by tuning these hyperparameters but not by adding additional parameters. This is because at the hyperparmeters already tuned for the model, if we add more, then we may lead to a situation where it would take the model more time to learn and we need to additionally tune the entire hyperparameter set again which leads to more work and increases the complexity of the model. We need to employ cross-validation techniques to tune these hyperparameters together and then only we can achieve better accuracy.

**Q4.** Is it always worth training for more epochs (while keeping all other hyperparameters fixed)?

**Ans.** No it is not always worth training the model for more epochs for fixed values of the hyperparamters. This is because this can result in overfitting. However in certain cases, we can train the model on more epochs until the resulting validation accuracy and training accuracy are almost equal. As soon as either of the accuracies starts dropping, the model can overfit and can lead to lesser accuracy. For smaller problems, increasing the epochs keeping the parameters fixed would lead to no harm. But it is not always worth doing. It should be decided based on the problem at hand.